### PR TITLE
paras: count upgrade delay from inclusion

### PR DESCRIPTION
**Pull request by slumber**
Resolves https://github.com/paritytech/polkadot/issues/4601

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3160103

File: runtime/parachains/src/paras/mod.rs [hunk 1]
```diff
@@ -367,16 +367,20 @@ impl TypeInfo for ParaKind {
...
 		id: ParaId,
-		/// The relay-chain block number that was used as the relay-parent for the parablock that
-		/// initiated the upgrade.
+		/// The relay-chain block number of **inclusion** of candidate that that initiated the upgrade.
+		///
+		/// It's important to count upgrade enactment delay from the inclusion of this candidate instead
+		/// of its relay parent -- in order to keep PVF available in case of chain reversions.
+		///
+		/// See https://github.com/paritytech/polkadot/issues/4601 for detailed explanation.
 		relay_parent_number: BlockNumber,
```
Comment (ordian) [inline runtime/parachains/src/paras/mod.rs:384]: can we rename this to `inclusion_parent_number` or `inclusion_block_number`?
```suggestion
		inclusion_parent_number: BlockNumber,
```

**Comment by slumber** [comment]
> Wouldn't nodes disputing candidate with relay parent 1939 fetch PVF from that block or some recent block and not inclusion block?

The problem is it won't be available. The code gets stored on candidate's inclusion, however the delay for its enactment is counted from candidate's relay parent. So in total we're off by `chain_availability_period` + (with async backing) `allowed_ancestry_len` blocks. In numbers, that would mean that we're off by approx 10 + 5 blocks.

> This seems like a partial solution to a very contrived scenario, so not sure how helpful this change is.

It is partial in a sense that `validation_upgrade_delay` is not deterministic value, or rather, it doesn't restrict chain from reverting past this number, but if we want to rely on it (i.e. assuming it is), we have to count from inclusion, otherwise (from the above paragraph) it's broken.

**Review by rphmeier — APPROVED** [review]
As mentioned in the original issue, this is strictly more correct now that code upgrades are handled with on-chain signals as opposed to coordinated block numbers. The `GoAheadSignal` infrastructure has been in place for over a year now, so all chains definitely make use of it.

**Comment by rphmeier** [comment]
Nit: I really don't like the term "inclusion parent", since the 'parent' part isn't important for this case. `included_at` or `included_in` makes more sense.

**Comment by slumber** [comment]
bot merge

**Comment by slumber** [comment]
bot merge

**Comment by slumber** [comment]
CI failure seems to be unrelated

**Comment by slumber** [comment]
bot merge

**Pull request merged**